src/main/GCNodeAllocator.cpp

```diff
@@ -0,0 +1,440 @@
+ * track all current allocations. This map is used to double-check
+ * each operation the GCNodeAllocator does to ensure it is consistent
+ * with the state recorded in the ALLOCATION_CHECK map.
+ */
+
+typedef std::map<void*, void*> allocation_map;
+
+/** Extra map for allocator sanity checking. */
+static allocation_map allocations;
+
+/** Adds an allocation to the extra allocation map. */
+static void add_to_allocation_map(void* allocation, std::size_t size) {
+  void* allocation_end = static_cast<char*>(allocation) + size;
+  allocations[allocation] = allocation_end;
+}
+
+/** Removes an allocation from the extra allocation map. */
+static void remove_from_allocation_map(void* allocation) {
+  allocations.erase(allocation);
+}
+
+/**
+ * Check if a pointer is in the extra allocation map.
+ * Returns the start pointer of the allocation if the pointer
+ * is in the map, otherwise returns null.
+ */
+static void* lookup_in_allocation_map(void* tentative_pointer) {
+  // Find the largest key less than or equal to tentative_pointer.
+  allocation_map::const_iterator next_allocation =
+      allocations.upper_bound(tentative_pointer);
+  if (next_allocation != allocations.begin()) {
+    allocation_map::const_iterator allocation = std::prev(next_allocation);
+
+    // Check that tentative_pointer is before the end of the allocation.
+    void* allocation_end = allocation->second;
+    if (tentative_pointer < allocation_end) {
+      return allocation->first;
+    }
+  }
+  return nullptr;
+}
+#endif
+
+#ifdef HAVE_ADDRESS_SANITIZER
+void* rho::GCNodeAllocator::offsetPointer(void* pointer, std::size_t bytes) {
+    return reinterpret_cast<void*>(reinterpret_cast<uintptr_t>(pointer) + bytes);
+}
+
+void rho::GCNodeAllocator::addToQuarantine(rho::FreeListNode* free_node, unsigned size_class) {
+  // Add an allocation to the quarantine and poison it.
+  static unsigned quarantine_size = 0;  // Separate from small object quarantine size.
+
+  size_t block_size = bytesFromSizeClass(size_class);
+  free_node->m_next = s_quarantine[size_class];
+  s_quarantine[size_class] = free_node;
+  ASAN_POISON_MEMORY_REGION(free_node, block_size);
+
+  // Clear the quarantine if it has grown too large.
+  quarantine_size += block_size;
+  if (quarantine_size > s_max_quarantine_size) {
+    quarantine_size = 0;
+    // Now we clear the quarantine.
+    // The quarantine contains both blocks in superblocks and large allocations.
+    // Remove all free nodes from quarantine and insert in corresponding freelist.
+    for (int i = 0; i < s_num_freelists; ++i) {
+      while (s_quarantine[i]) {
+        FreeListNode* quarantined_node = s_quarantine[i];
+
+        // Unpoison the quarantined allocation so we can link it into a freelist.
+        ASAN_UNPOISON_MEMORY_REGION(quarantined_node, sizeof(FreeListNode));
+        s_quarantine[i] = quarantined_node->m_next;
+        quarantined_node->m_next = s_freelists[i];
+        s_freelists[i] = quarantined_node;
```

Comment:
You can completely unroll this loop if you have the last element in the quarantine (or loop to find it):

```
last_element_in_quarantine->m_next = s_freelists[i];
s_freelists[i] = s_quarantine[i];
s_quarantine[i] = nullptr;
```